### DISCUSSION TITLE
How does `_AppStorage` interoperate with SwiftUI `AppStorage`?

### DISCUSSION THREAD

**Discussion by Zeta611** [tgrapperon/swift-dependencies-additions#32] [Q&A] [answered]
The following does not properly inject values:
```swift
withDependencies {
  $0.userDefaults = .ephemeral()
  $0.userDefaults.set("hello", forKey: "myKey")
} operation: {
  MyView()
  }
```
where `MyView` has
```swift
@AppStorage("myKey") var myKey = "init"
```
It works when I replace `@AppStorage("myKey")` with `@Dependency.AppStorage("myKey")` though.

However, `README` says that they should interoperate when they have the same keys, and yet it doesn't seem to work the way I expected.
How are they supposed to interoperate?

Thank you!

**Comment by tgrapperon** [comment, marked as answer]
Hey @Zeta611! Sorry if this wasn't clear: they interoperate only with the default live implementation. In your case, you're using the `ephemeral` storage, but the SwiftUI variant doesn't know anything about that. They interoperate when you don't override `userDefault` storage (or use the same `UserDefaults` instance), because they're writing to the same place, in the same fashion, and they both observe changes to this storage. It means that you can create a `Settings` model that writes to `@Dependency.AppStorage`, and pop this preference in any `View` elsewhere in the hierarchy using `SwiftUI`'s `@AppStorage` for example.

It is partly working here because you defined your `View` in a `withDependency` block, but `@Dependency.AppStorage` will not invalidate the view if its value change. In a general manner, I wouldn't use `@Dependency` stuff in view, excepted very locally for previews (but as I said earlier, ephemeral() app storage wouldn't interoperate with `SwiftUI`'s `AppStorage`).

I'll think about rephrasing this to make it clearer. I'll also think if I could make `ephemeral` work in a `View`, but I'm not sure this is possible.

**Comment by Zeta611** [comment]
Thank you for the clarification!
I have one more follow up question to clarify: If I don't explicitly set `$0.userDefaults = .liveValue`, `@Dependency.AppStorage` does not interoperate with `@AppStorage` in a preview. So is `.previewValue` used automatically in the context of a `PreviewProvider`?

**Reply by tgrapperon** [reply]
Yes `previewValue` is automatically used in a `PreviewProvider`, that's its main feature. 
As previews are considered discardable, I've opted to default `userDefault` to `ephemeral` for them, so you always know how with what you're starting with. I may add an additional closure that would allow to define initial values though.

**Reply by Zeta611** [reply]
Thank you very much for the prompt reply!
